pkg/ccl/backupccl/restore_old_versions_test.go

```diff
@@ -248,3 +250,194 @@ func restoreOldVersionClusterTest(exportDir string) func(t *testing.T) {
+	rng, _ := randutil.NewPseudoRand()
+
+	sqlDB.Exec(t, `CREATE TABLE data.jsontest (id INT PRIMARY KEY, j JSONB)`)
+	sqlDB.Exec(t, `INSERT INTO data.jsontest VALUES (1, '{"a": "a", "b": "b"}'), (2, '{"c": "c", "d":"d"}')`)
+
+	sqlDB.Exec(t, `CREATE TABLE data.geotest (id INT PRIMARY KEY, p geometry(point))`)
+	sqlDB.Exec(t, `INSERT INTO data.geotest VALUES (1, 'POINT(1.0 1.0)'), (2, 'POINT(2.0 2.0)')`)
+
+	var backupDirs []string
+	var checksums []uint32
+	{
+		for backupNum := 0; backupNum < numBackups; backupNum++ {
+			// In the following, windowSize is `w` and offset is `o`. The first
+			// mutation creates accounts with id [w,3w). Every mutation after
+			// that deletes everything less than o, leaves [o, o+w) unchanged,
+			// mutates [o+w,o+2w), and inserts [o+2w,o+3w).
+			offset := windowSize * backupNum
+			var buf bytes.Buffer
+			fmt.Fprintf(&buf, `DELETE FROM data.bank WHERE id < %d; `, offset)
+			buf.WriteString(`UPSERT INTO data.bank VALUES `)
+			for j := 0; j < windowSize*2; j++ {
+				if j != 0 {
+					buf.WriteRune(',')
+				}
+				id := offset + windowSize + j
+				payload := randutil.RandBytes(rng, backupRestoreRowPayloadSize)
+				fmt.Fprintf(&buf, `(%d, %d, '%s')`, id, backupNum, payload)
+			}
+			sqlDB.Exec(t, buf.String())
+			createErr := make(chan error)
+			go func() {
+				defer close(createErr)
+				var stmt string
+				switch backupNum % 3 {
+				case 0:
+					stmt = fmt.Sprintf(`CREATE INDEX test_idx_%d ON data.bank (balance)`, backupNum+1)
+				case 1:
+					stmt = fmt.Sprintf(`CREATE INDEX test_idx_%d ON data.jsontest USING GIN(j)`, backupNum+1)
+				case 2:
+					stmt = fmt.Sprintf(`CREATE INDEX test_idx_%d ON data.geotest USING GIST(p)`, backupNum+1)
+				}
+				t.Log(stmt)
+				_, err := sqlDB.DB.ExecContext(ctx, stmt)
+				createErr <- err
+			}()
+			select {
+			case <-backfillWaiting:
+			case err := <-createErr:
+				t.Fatal(err)
+			}
+			checksums = append(checksums, checksumBankPayload(t, sqlDB))
+
+			backupDir := fmt.Sprintf("nodelocal://0/%d", backupNum)
+			var from string
+			if backupNum > 0 {
+				from = fmt.Sprintf(` INCREMENTAL FROM %s`, strings.Join(backupDirs, `,`))
+			}
+			sqlDB.Exec(t, fmt.Sprintf(`BACKUP TO '%s' %s`, backupDir, from))
+			blockBackfill <- struct{}{}
+			require.NoError(t, <-createErr)
+
+			backupDirs = append(backupDirs, fmt.Sprintf(`'%s'`, backupDir))
+		}
+
+		// Test a regression in RESTORE where the batch end key was not
+		// being set correctly in Import: make an incremental backup such that
+		// the greatest key in the diff is less than the previous backups.
+		sqlDB.Exec(t, `INSERT INTO data.bank VALUES (0, -1, 'final')`)
+		checksums = append(checksums, checksumBankPayload(t, sqlDB))
+		sqlDB.Exec(t, fmt.Sprintf(`BACKUP TO '%s' %s`,
+			"nodelocal://0/final", fmt.Sprintf(` INCREMENTAL FROM %s`, strings.Join(backupDirs, `,`)),
+		))
+		backupDirs = append(backupDirs, `'nodelocal://0/final'`)
+	}
+	os.Rename(dir, path/to/testdata)
```

Comment:
nit if you're going to push again: although this will fail anyway, I think this is missing a closing brace.

Reply:
heh, indeed. done.